frontends/openssl-provider/common.c

```diff
@@ -0,0 +1,406 @@
+}
+
+static int esdm_rand_instantiate(void *ctx __unused,
+                                 unsigned int strength __unused,
+                                 int prediction_resistance __unused,
+                                 const unsigned char *pstr __unused,
+                                 size_t pstr_len __unused,
+                                 const OSSL_PARAM params[] __unused)
+{
+  return 1;
+}
+
+static int esdm_rand_uninstantiate(void *ctx __unused)
+{
+  return 1;
+}
+
+static int esdm_rand_generate(void *ctx __unused, unsigned char *out,
+                              size_t outlen, unsigned int strength __unused,
+                              int prediction_resistance,
+                              const unsigned char *addin, size_t addin_len)
+{
+  ssize_t ret;
+
+  if (!out)
+    goto err;
+
+  if (addin && addin_len > 0) {
+    esdm_invoke(esdm_rpcc_write_data(addin, addin_len));
+    if (ret != 0)
+      goto err;
+  }
+
+  if (prediction_resistance) {
+    esdm_invoke(esdm_rpcc_get_random_bytes_pr(out, outlen));
+  } else {
+    esdm_invoke(esdm_rpcc_get_random_bytes_full(out, outlen));
+  }
+  if (ret != (ssize_t)outlen)
+    goto err;
+
+  return 1;
+
+err:
+  return 0;
+}
+
+static int esdm_rand_reseed(void *ctx __unused,
+                            int prediction_resistance __unused,
+                            const unsigned char *ent, size_t ent_len,
+                            const unsigned char *addin, size_t addin_len)
+{
+  ssize_t ret;
+
+  /* unaccounted writing of additional data does no harm */
+  if (ent && ent_len > 0)
+    esdm_invoke(esdm_rpcc_write_data(ent, ent_len));
+  if (addin && addin_len > 0)
+    esdm_invoke(esdm_rpcc_write_data(addin, addin_len));
+
+  return 1;
+}
+
+static size_t esdm_rand_nonce(void *ctx __unused, unsigned char *out,
+                              unsigned int outlen, size_t min_noncelen __unused,
+                              size_t max_noncelen __unused)
+{
+  ssize_t ret;
+
+  if (out == NULL)
+    return outlen;
+
+  esdm_invoke(esdm_rpcc_get_random_bytes_full(out, outlen));
+
+  if (ret == (ssize_t)outlen)
+    return outlen;
+  else
+    return 0;
+}
+
+static size_t esdm_rand_get_seed(void *ctx __unused, unsigned char **buffer,
+                                 int entropy, size_t min_len __unused,
+                                 size_t max_len __unused,
+                                 int prediction_resistance __unused,
+                                 const unsigned char *addin, size_t addin_len)
+{
+  struct esdm_seed_buffer {
+    uint64_t len;
+    uint64_t entropyBits;
+    uint8_t buf[2048] /* should be large enough for entropy from all active
+                         sources */
+  } __attribute__((__packed__));
+  size_t buffer_size = sizeof(struct esdm_seed_buffer);
+  struct esdm_seed_buffer *seed_buf;
+  ssize_t ret;
+
+  *buffer = OPENSSL_secure_zalloc(sizeof(struct esdm_seed_buffer));
```

Comment:
Shouldn't the returned data only hold the seed_buf and not the full structure? The caller may have no notion about the seed_buf structure. But I admit that it does not hurt to also return the additional information.

Reply:
I changed that accordingly. I just thought of the header values as additional data, that does not harm if later cryptographically hashed (and no entropy is accounted for it).